tiny-firmware/firmware/fsm.c

```diff
@@ -664,3 +665,55 @@ void fsm_msgBitcoinTxAck(BitcoinTxAck *msg) {
     layoutHome();
     return;
 }
+
+static void get_root_node_callback(uint32_t iter, uint32_t total) {
+  usbSleep(1);
+  layoutProgress("Waking up", 1000 * iter / total);
+}
+
+int fsm_getBitcoinBIP39_Seed(uint8_t seed[]){
+
+     char mnemonic[MAX_MNEMONIC_LEN + 1];
+     const char *mnemo = storage_getFullSeed();
+
+     strcpy(mnemonic, mnemo);
+
+     if(!mnemonic_check(mnemonic)){
+       return 1;
+     }
+
+     char oldTiny = usbTiny(1);
+     mnemonic_to_seed(mnemonic, "", seed, get_root_node_callback);
+     memzero(mnemonic, sizeof(mnemonic));
+     usbTiny(oldTiny);
+
+     return 0;
+}
+
+HDNode* fsm_getDerivedNode(const char *curve){
+
+  static CONFIDENTIAL HDNode node;
+  static uint8_t CONFIDENTIAL BIP39Seed[256];
+
+  fsm_getBitcoinBIP39_Seed(BIP39Seed);
+
+  // Chain m
+  hdnode_from_seed(BIP39Seed, 64, curve, &node);
+
+  // if (!address_n || address_n_count == 0) {
+  //   return &node;
+  // }
+
+  // Chain m/0'
+  if(hdnode_private_ckd_prime(&node, 0) == 0){
+    return 0;
+  }
+
+  // Chain m/0'/0
+  if (hdnode_private_ckd(&node, 0) == 0){
+
+      return 0;
+  }
+
+  return &node;
+}
```

Comment:
This functions should not be in this file, as they are not part of communication between device and computer, I think it's better to put them somewhere in skycoin-crypto

Reply:
Is not possible, to move it to skycoin_crypto, because those functions used `storage.h` and `layout.h`, which are part of the firmware. Creating a new file in `firmware` directory and moving those functions to this file is probably the best solution.